plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -394,6 +393,157 @@ private S3AsyncClient createS3AsyncClient(AwsCredentialsProvider credentialsProv
+        private long fileOffset;
+
+        private byte[] buffer;
+        private int bufferPosition;
+        private int bufferFill;
+
+        private int sliceSize = -1;
+        private ListenableFuture<Void> inProgressReadFuture = immediateVoidFuture();
+
+        public S3ExchangeStorageReader(
+                S3AsyncClient s3AsyncClient,
+                Iterator<ExchangeSourceFile> exchangeSourceFiles,
+                int partSize,
+                int maxPageStorageSize)
+        {
+            this.s3AsyncClient = requireNonNull(s3AsyncClient, "s3AsyncClient is null");
+            this.exchangeSourceFiles = requireNonNull(exchangeSourceFiles, "exchangeSourceFiles is null");
+            this.partSize = partSize;
+            // Make sure buffer can accommodate at least one complete Slice, and keep reads aligned to part boundaries
+            this.bufferSize = maxPageStorageSize + partSize;
+
+            // This class is expected to be used in a single-threaded manner, therefore assignment to non-final fields
+            // doesn't constitute unsafe publication
+            checkArgument(exchangeSourceFiles.hasNext(), "No exchange source file to read from");
+            this.currentFile = exchangeSourceFiles.next();
+            fillBuffer();
+        }
+
+        @Override
+        public Slice read()
+                throws IOException
+        {
+            if (currentFile == null) {
+                return null;
+            }
+
+            try {
+                getFutureValue(inProgressReadFuture);
+            }
+            catch (RuntimeException e) {
+                Throwable cause = e.getCause() == null ? e : e.getCause();
+                throw new IOException(cause);
+            }
+
+            verify(bufferFill - bufferPosition > Integer.BYTES, "Buffer content should contain at least one complete slice");
+            if (sliceSize < 0) {
+                sliceSize = ByteBuffer.wrap(buffer, bufferPosition, Integer.BYTES).order(LITTLE_ENDIAN).getInt();
+            }
+            bufferPosition += Integer.BYTES;
+
+            verify(0 < sliceSize && sliceSize <= bufferFill - bufferPosition,
+                    "Buffer content should contain at least one complete slice");
+            Slice slice = Slices.wrappedBuffer(buffer, bufferPosition, sliceSize);
+            bufferPosition += sliceSize;
+
+            if (bufferFill - bufferPosition > Integer.BYTES) {
+                sliceSize = ByteBuffer.wrap(buffer, bufferPosition, Integer.BYTES).order(LITTLE_ENDIAN).getInt();
+                if (bufferFill - bufferPosition - Integer.BYTES < sliceSize) {
+                    fillBuffer();
+                }
+            }
+            else {
+                sliceSize = -1;
+                fillBuffer();
+            }
+
+            return slice;
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            return nonCancellationPropagating(inProgressReadFuture);
```

Comment:
Given that now we want S3ExchangeStorageReader to be thread-safe, then we will still keep `nonCancellationPropagating `

Reply:
While from the point of view of this interface it might be better to keep it `nonCancellationPropagating` I'm thinking if it is better to avoid it from practicality reasons and only keep the `nonCancellationPropagating` at the `ExchangeSource` interface level (to avoid creating unnecessary objects and attaching unnecessary listeners)